### ISSUE TITLE
go routine leak in scan.go?

### ISSUE THREAD

**Issue by miekg** [miekg/dns#568]
from https://github.com/coredns/coredns/issues/1233

~~~
goroutine profile: total 27
9 @ 0x3e6c8 0x3e790 0x14124 0x13ea8 0x224060 0x6fd2c
#	0x22405f	github.com/miekg/dns.zlexer+0x347 /home/miek/g/src/github.com/miekg/dns/scan.go:602
~~~

This is code that calls `NewRR` which starts a goroutine to parse, maybe that one can be made not to return?

Highly speculative atm, btw.

**Comment by miekg** [comment]
Number of leaks has 1:1 mapping with seeing the following record:
`mplus.ims.vodafone.com. 1866 NAPTR 10 100 S SIP+D2U  _sip._udp.mplus.ims.vodafone.com.`

**Comment by miekg** [comment]
According to the spec S and SIP+D2U need to be enclosed in quotes, see https://www.ietf.org/rfc/rfc2915.txt

https://dns.google.com/resolve?name=mplus.ims.vodafone.com.&type=NAPTR
{"Status": 0,"TC": false,"RD": true,"RA": true,"AD": false,"CD": false,"Question":[ {"name": "mplus.ims.vodafone.com.","type": 35}],"Answer":[ {"name": "mplus.ims.vodafone.com.","type": 35,"TTL": 1985,"data": "10 100 S SIPS+D2T  _sips._tcp.mplus.ims.vodafone.com."},{"name": "mplus.ims.vodafone.com.","type": 35,"TTL": 1985,"data": "10 100 S SIP+D2U  _sip._udp.mplus.ims.vodafone.com."},{"name": "mplus.ims.vodafone.com.","type": 35,"TTL": 1985,"data": "10 100 S SIP+D2T  _sip._tcp.mplus.ims.vodafone.com."}]}

**Comment by miekg** [comment]
So I think this is happening:
We crap out of scan_rr.go because the record is wrong: https://github.com/miekg/dns/blob/master/scan_rr.go#L690
But the `chan lex` is not closed (that happens in the zlexer function with a `defer`), so that stays alive, and the scanText also doesn't see a error, so from inside `zlexer` everything looks OK - hence goroutine leak.

This is true for all RRs btw, just showed up in this NAPTR.

**Comment by miekg** [comment]
Closed via #570
